www/blog/2022-11-23-ts-perf.md

```diff
@@ -0,0 +1,221 @@
+- Having a large initial load time
+- If the TypeScript language server takes a long time to respond to changes
+
+The last point is one that the tRPC must pay attention to the most. You **never** want to your developers to have to wait for the language server to update after a change. This is where tRPC **must** maintain performance so that you can enjoy great DX.
+
+## TypeScript library optimization strategies
+
+Perhaps the most important thing to understand about optimizing TypeScript is the fact that it is duck typed. This characteristic of TS comes from the historical background of being a superset of JavaScript, where there are many...um...flexible, creative ways to write code that...erm...works. Supersetting types that work for the open-ended language of JavaScript naturally gives way to TypeScript being...uh...inventive. This means that there are often many ways to accomplish the same goal.
+
+Where this becomes important is that library authors need to understand the interfaces of their library against application code. The types exposed to developers need to be strict enough to maintain a proper program - but they also need to be performant. Type-checking is great - but when can we get the same result from doing less work?
+
+Let's look at an example where we have an object creating a dictionary of values for setting and using colors in our application.
+
+```ts
+export const colorHexMap = {
+  red: '#ff0000',
+  green: '#00ff00',
+  blue: '#0000ff',
+  // ...1000 more colors
+} as const;
+```
+
+In this example, we have a [const assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions). `as const` tells the TypeScript compiler that a type is not allowed to be widened to a more generalized type. For instance, the type of `colorHexMap` will be known to be the **specific** set of keys **and** their **specific** set of values. `colorHexMap.red` is not of type `string` - but the string literal of `"#ff0000"`.
+
+Wow, that sounds pretty cool. TS will tell me exactly what the value is of `red`!
+
+But the tightness of these types has a cost - compiler performance. Let's consider the same example but with a slightly looser type:
+
+```ts
+export const colorHexMap: Record<string, string> = {
+  red: '#ff0000',
+  green: '#00ff00',
+  blue: '#0000ff',
+  // ...1000 more colors
+};
+```
+
+A `Record<string, string>` is _vastly_ simpler for the TypeScript compiler to handle. We're still appropriately defining what a `colorHexMap` looks like as a type but we will get much better performance.
+
+Clearly, there is a tradeoff here and we need to be mindful of what our types represent in our code. This is where great engineers shine: understanding tradeoffs and how those tradeoffs affect programs, other developers, and end users. In the example above, will it be possible for the application to run into severe errors because of the loosening of the type? Is the performance gain worth it?
```

Comment:
I get it and I like it but I feel like this sentence might be weird coming from me: "This is where great engineers shine: understanding tradeoffs and how those tradeoffs affect programs, other developers, and end users." I don't think I have enough years on other people to speak like this

Reply:
Hahaha, that's extremely fair.